src/java.base/share/classes/java/lang/runtime/SwitchBootstraps.java

```diff
@@ -432,26 +451,90 @@ record Element(Label target, Label next, Object caseLabel) {}
+                                    } else {
+                                        Label compare = cb.newLabel();
+                                        cb.invokevirtual(ConstantDescs.CD_Number,
+                                                         "intValue",
+                                                         MethodTypeDesc.of(ConstantDescs.CD_int));
+                                        cb.goto_(compare);
+                                        cb.labelBinding(notNumber);
+                                        cb.aload(0);
+                                        cb.instanceof_(ConstantDescs.CD_Character);
+                                        cb.ifeq(next);
+                                        cb.aload(0);
+                                        cb.checkcast(ConstantDescs.CD_Character);
+                                        cb.invokevirtual(ConstantDescs.CD_Character,
+                                                         "charValue",
+                                                         MethodTypeDesc.of(ConstantDescs.CD_char));
+                                        cb.labelBinding(compare);
+                                    }
+
+                                    TypePairs typePair = TypePairs.of(Wrapper.asPrimitiveType(selectorType), classLabel);
+                                    String methodName = typePairToName.get(typePair);
+                                    cb.invokestatic(ExactConversionsSupport.class.describeConstable().orElseThrow(),
+                                                    methodName,
+                                                    MethodTypeDesc.of(ConstantDescs.CD_boolean, typePair.from.describeConstable().orElseThrow()));
+                                    cb.ifeq(next);
+                                }
                             } else {
-                                cb.aload(3);
-                                cb.constantInstruction(extraClassLabels.size());
-                                cb.invokeinterface(ConstantDescs.CD_List,
-                                                   "get",
-                                                   MethodTypeDesc.of(ConstantDescs.CD_Object,
-                                                                     ConstantDescs.CD_int));
-                                cb.checkcast(ConstantDescs.CD_Class);
-                                cb.aload(0);
-                                cb.invokevirtual(ConstantDescs.CD_Class,
-                                                 "isInstance",
-                                                 MethodTypeDesc.of(ConstantDescs.CD_boolean,
-                                                                   ConstantDescs.CD_Object));
-                                cb.ifeq(next);
-                                extraClassLabels.add(classLabel);
+                                Optional<ClassDesc> classLabelConstableOpt = classLabel.describeConstable();
```

Comment:
One comment (for future work): we have a bunch of constants that are known by the time we generate this code (at runtime). But we're still serializing them into the constant pool, and then deserialize them at runtime, which is needless work. Hidden classes support injection of already resolved constants into the constant pool - see:

https://download.java.net/java/early_access/jdk22/docs/api/java.base/java/lang/invoke/MethodHandles.Lookup.html#defineHiddenClassWithClassData(byte%5B%5D,java.lang.Object,boolean,java.lang.invoke.MethodHandles.Lookup.ClassOption...)

Basically, the idea is that you can pass the entire `labels` array as a "constant data" which is stored in the constant pool as a pre-resolved constant. You can then load elements from this array using ready-made method handles:

https://download.java.net/java/early_access/jdk22/docs/api/java.base/java/lang/invoke/MethodHandles.html#classDataAt(java.lang.invoke.MethodHandles.Lookup,java.lang.String,java.lang.Class,int)